torch/_inductor/memory.py

```diff
@@ -427,13 +438,47 @@ class BufferInfo(TypedDict):
             if buf_info[buf]["outdegree"] == 0:
                 node_info[node]["memory_to_free"] += buf.mpi_buffer.size_free
 
+    # get list of collective nodes
+    collective_indices_remaining = [
+        node.mpi_node.index
+        for node in nodes
+        if (node.node)
+        and (is_collective(node.node) or is_torchrec_collective(node.node))
+    ]
+    next_collective_index = (
+        collective_indices_remaining.pop(0)
+        if collective_indices_remaining
+        else len(nodes)
+    )
+
     # schedule nodes one at a time
     schedule: List[BaseSchedulerNode] = []
     num_iters: int = 0
     while num_iters < len(nodes) and nodes_to_schedule:
+        # if there is a collective node in the graph, we restrict reordering to nodes
+        # that are before the next collective node
+        if collective_indices_remaining:
+            if num_iters == next_collective_index:
+                nodes_to_schedule_allowed = OrderedSet([nodes[next_collective_index]])
+                next_collective_index = (
+                    collective_indices_remaining.pop(0)
+                    if collective_indices_remaining
+                    else len(nodes)
+                )
+            else:
+                nodes_to_schedule_allowed = OrderedSet(
+                    [
+                        node
+                        for node in nodes_to_schedule
+                        if node.mpi_node.index < next_collective_index
+                    ]
+                )
+        else:
+            nodes_to_schedule_allowed = nodes_to_schedule
```

Comment:
Curious what would be the difference between this approach vs. turning on `comms.decide_global_ordering_of_comms` unconditionally in scheduler.py, and have that pass also check `is_torchrec_collective()`?

Reply:
There is no significant difference in memory utilization or qps. But for correctness:
1. Here is a baseline [run](https://www.internalfb.com/mast/job/aps-xuanzh-e9e01b1d29) (on main branch) which has NCCL timeout as expected 
2. Here is a [run](https://www.internalfb.com/mast/job/aps-xuanzh-71d91a2d22) with no change to the memory pass but the global comm order includes torchrec comm ops, and it runs without problems. https://github.com/pytorch/pytorch/pull/141968

Maybe it is better to do the fix this in the second way?

Reply:
Thanks for the info! yes doing it the second way sounds good, since iiuc they fundamentally are doing the same thing